src/systems/world_exporter/WorldExporter.cc

```diff
@@ -0,0 +1,382 @@
+    std::map<Entity, std::pair<math::Pose3d, Entity>> poses;
+    std::vector<math::Matrix4d> subMeshMatrix;
+    common::ColladaExporter exporter;
+
+  private: math::Pose3d GetPose(Entity parent_, math::Pose3d currentPose)
+  {
+    std::stack<math::Pose3d> posesStack;
+    std::pair<math::Pose3d, Entity> curr;
+    math::Pose3d localPose;
+
+    posesStack.push(currentPose);
+    while (world != parent_)
+    {
+      curr = this->poses.at(parent_);
+      posesStack.push(curr.first);
+      parent_ = curr.second;
+    }
+    while (!posesStack.empty())
+    {
+      localPose = localPose * posesStack.top();
+      posesStack.pop();
+    }
+    igndbg << ">>>>  - POSE : " << localPose <<std::endl;
+
+    return localPose;
+  }
+
+  public: void Export(const EntityComponentManager &_ecm)
+  {
+    if (this->rendered) return;
+
+    _ecm.Each<components::World, components::Name>(
+      [&](const Entity & _entity,
+        const components::World *,
+        const components::Name * _name)->bool
+    {
+      world = _entity;
+      this->worldMesh.SetName(_name->Data());
+      return true;
+    });
+
+    _ecm.Each<components::Model, components::Pose,
+            components::ParentEntity>(
+      [&](const Entity &_entity,
+          const components::Model *,
+          const components::Pose *_pose,
+          const components::ParentEntity *_parent)->bool
+    {
+      poses.insert(std::pair<Entity, std::pair<math::Pose3d, Entity>>(
+        _entity, std::pair<math::Pose3d, Entity>(
+           _pose->Data(), _parent->Data()) ));
+      return true;
+    });
+
+    _ecm.Each<components::Link, components::Pose,
+            components::ParentEntity>(
+      [&](const Entity &_entity,
+          const components::Link *,
+          const components::Pose *_pose,
+          const components::ParentEntity *_parent)->bool
+    {
+      poses.insert(std::pair<Entity, std::pair<math::Pose3d, Entity>>(
+        _entity, std::pair<math::Pose3d, Entity>(
+           _pose->Data(), _parent->Data()) ));
+      return true;
+    });
+
+
+    _ecm.Each<components::Visual,
+            components::Name,
+            components::Pose,
+            components::Geometry,
+            components::Transparency,
+            components::ParentEntity>(
+    [&](const ignition::gazebo::Entity &_entity,
+        const components::Visual *,
+        const components::Name *_name,
+        const components::Pose *_pose,
+        const components::Geometry *_geom,
+        const components::Transparency *_transparency,
+        const components::ParentEntity *_parent)->bool
+    {
+      std::string name = _name->Data().empty() ? std::to_string(_entity) :
+          _name->Data();
+
+      igndbg << "====================== Starting: " << name << " - "
+        << _entity << "  ====================== " << std::endl;
+
+      math::Pose3d localPose = this->GetPose(_parent->Data(), _pose->Data());
+
+      common::MaterialPtr mat(new common::Material());
```

Comment:
This is slightly more efficient:

```suggestion
      common::MaterialPtr mat = std::make_shared<common::Material>();
```